src/fp/list.nim

```diff
@@ -114,6 +114,24 @@ proc foldRightF*[T, U](xs: List[T], z: () -> U, f: (T, () -> U) -> U): U =
   if xs.isEmpty: z()
   else: f(xs.head, () => xs.tail.foldRightF(z, f))
 
+proc unfoldLeft*[T](f: T -> Option[(T, T)], x:T): List[T] {. inline .}=
+    ## Build a List from the left from function f: T -> Option(T,T) and a seed of type T
+    result = Nil[T]()
+    var a, b = x
+
+    while f(a).isDefined():
```

Comment:
This way `f(a)` is executed twice per returned element. Not only is this a performance hit, it also relies on `f(a)` being referentially transparent.